api/src/services/xml_generation/header_generator.py

```diff
@@ -0,0 +1,117 @@
+        if hash_value:
+            hash_element = lxml_etree.SubElement(root, f"{{{HEADER_NAMESPACES['glob']}}}HashValue")
+            hash_element.set(f"{{{HEADER_NAMESPACES['glob']}}}hashAlgorithm", "SHA-1")
+            hash_element.text = hash_value
+
+        # Add header fields (only if not None)
+        self._add_header_field(root, "AgencyName", self._get_agency_name())
+        self._add_header_field(root, "OpportunityID", self.opportunity.opportunity_number)
+        self._add_header_field(root, "OpportunityTitle", self.opportunity.opportunity_title)
+        self._add_header_field(root, "CompetitionID", self.competition.public_competition_id)
+        self._add_header_field(
+            root, "OpeningDate", self._format_date(self.competition.opening_date)
+        )
+        self._add_header_field(
+            root, "ClosingDate", self._format_date(self.competition.closing_date)
+        )
+        self._add_header_field(root, "SubmissionTitle", self.application.application_name)
+        self._add_header_field(root, "CFDANumber", self._get_cfda_number())
+
+        # Generate XML string
+        if pretty_print:
+            xml_bytes = lxml_etree.tostring(
+                root, encoding="utf-8", xml_declaration=True, pretty_print=True
+            )
+        else:
+            xml_bytes = lxml_etree.tostring(root, encoding="utf-8", xml_declaration=True)
+
+        return xml_bytes.decode("utf-8").strip()
+
+    def _add_header_field(
+        self, parent: lxml_etree.Element, field_name: str, value: str | None
+    ) -> None:
+        if value is not None:
+            element = lxml_etree.SubElement(
+                parent, f"{{{HEADER_NAMESPACES['header']}}}{field_name}"
+            )
+            element.text = str(value)
+
+    def _get_agency_name(self) -> str | None:
+        if self.opportunity.agency_name:
+            return self.opportunity.agency_name
+        return self.opportunity.agency_code
+
+    def _get_cfda_number(self) -> str | None:
+        if self.competition.opportunity_assistance_listing:
+            return self.competition.opportunity_assistance_listing.assistance_listing_number
+        return None
+
+    def _format_date(self, date_value: date | None) -> str | None:
+        if date_value is None:
+            return None
+        return date_value.isoformat()
+
+    # TODO: this is likely going to be changed based on additional research into how the hash value is calculated
+    def _calculate_hash_value(self) -> str | None:
+        # Create a string from key application data
+        hash_data_parts = [
+            str(self.application.application_id) if self.application.application_id else "",
+            self.application.application_name or "",
+            self.opportunity.opportunity_number or "",
+            str(self.competition.competition_id) if self.competition.competition_id else "",
+        ]
+        hash_data = "|".join(hash_data_parts)
```

Comment:
I couldn't confirm after digging through the code for a while, but my best guess at what the hash value is in grants.gov is either a Java class object hash OR some sort of hash of an XML object since it seems like it sometimes gets set upstream of the java logic.

I suppose one way we could try to get it to match would be to put the same data into XML and see if we can produce the same result, but at the same time.. if it's such an obtuse concept, we might just be fine putting whatever in it.

That's all to say, this is fine.